File: 189.cpp
// https://leetcode.com/problems/rotate-array/
class Solution {
 public:
  void rotate(vector<int>& nums, int k) {
    int sz = nums.size();
    k %= sz;

    auto it = std::next(std::begin(nums), sz - k);
    std::reverse(std::begin(nums), it);
    std::reverse(it, std::end(nums));
    std::reverse(std::begin(nums), std::end(nums));
  }
};

File: golang/189.go
package main

func reverse(a []int) {
	for i := len(a)/2 - 1; i >= 0; i-- {
		opp := len(a) - 1 - i
		a[i], a[opp] = a[opp], a[i]
	}
}

func rotate(nums []int, k int) {
	l := len(nums)
	k = k % l

	if k == 0 {
		return
	}

	k = l - k

	reverse(nums[:k])
	reverse(nums[k:])
	reverse(nums)
}
